File: content/problems/2516-take-k-of-each-character-from-left-and-right/solutions/longest-middle-window-with-count-limits/Solution.java
class Solution {
    public int takeCharacters(String s,int k){
        int[] total=new int[3];
        for(char character:s.toCharArray())total[character-'a']++;
        for(int count:total)if(count<k)return -1;
        int[] window=new int[3];
        int left=0,maximum=0;
        for(int right=0;right<s.length();right++){
            int index=s.charAt(right)-'a';
            window[index]++;
            while(window[index]>total[index]-k)window[s.charAt(left++)-'a']--;
            maximum=Math.max(maximum,right-left+1);
        }
        return s.length()-maximum;
    }
}

File: content/problems/2516-take-k-of-each-character-from-left-and-right/solutions/longest-middle-window-with-count-limits/Solution.kt
class Solution {
    fun takeCharacters(s:String,k:Int):Int{
        val total=IntArray(3);
        for(character in s)total[character-'a']++;
        if(total.any{
            it<k
        })return -1;
        val window=IntArray(3);
        var left=0;
        var maximum=0;
        for(right in s.indices){
            val index=s[right]-'a';
            window[index]++;
            while(window[index]>total[index]-k)window[s[left++]-'a']--;
            maximum=maxOf(maximum,right-left+1)
        };
        return s.length-maximum
    }
}
